staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,998 @@
+type DeviceRequest struct {
+	// By referencing a DeviceClass, a request inherits additional
+	// configuration parameters and selectors.
+	//
+	// A class is required. Which classes are available depends on the cluster.
+	//
+	// Administrators may use this to restrict which devices may get
+	// requested by only installing classes with selectors for permitted
+	// devices. If users are free to request anything without restrictions,
+	// then an empty class called "none" can get created to permit
+	// `deviceClassName: none`.
+	//
+	// +required
+	DeviceClassName string `json:"deviceClassName" protobuf:"bytes,1,name=deviceClassName"`
+
+	// Each selector must be satisfied by a device which is requested.
+	//
+	// +optional
+	// +listType=atomic
+	Selectors []Selector `json:"selectors,omitempty" protobuf:"bytes,2,name=selectors"`
+
+	// The count mode together with, for some modes, additional fields
+	// determines how many devices to allocate for the request.
+	//
+	// The default if unset is exactly one device:
+	//     countMode: Exact
+	//     count: 1
+	//
+	// "countMode: All" asks for all devices matching the selectors.
+	// Allocation fails if not all of them are available, unless admin
+	// access is requested. Admin access is granted also for
+	// devices which are in use.
+	//
+	// More modes may get added in the future.
+	//
+	// +default
+	CountMode string `json:"countMode,omitempty" protobuf:"bytes,3,opt,name=countMode"`
+
+	// Count is used only when the count mode is "Exact". Must be larger than zero.
+	//
+	// +optional
+	Count *int64 `json:"count,omitempty" protobuf:"bytes,4,opt,name=count"`
+
+	// AdminAccess indicates that this is a claim for administrative access
+	// to the device(s). Claims with AdminAccess are expected to be used for
+	// monitoring or other management services for a device.  They ignore
+	// all ordinary claims to the device with respect to access modes and
+	// any resource allocations. Ability to request this kind of access is
+	// controlled via ResourceQuota in the resource.k8s.io API.
+	//
+	// Default is false.
+	//
+	// +optional
+	AdminAccess *bool `json:"adminAccess,omitempty" protobuf:"bytes,5,opt,name=adminAccess"`
+}
+
+// Valid [DeviceRequest.CountMode] values.
+const (
+	CountModeExact = "Exact"
+	CountModeAll   = "All"
+)
+
+// Exactly one field must be set.
+type Selector struct {
+	// CEL contains a CEL expression for selecting a device.
+	//
+	// +required
+	CEL *CELSelector `json:"cel,omitempty" protobuf:"bytes,1,opt,name=cel"`
+}
+
+// CELSelector contains a CEL expression for selecting a device.
+type CELSelector struct {
+	// This CEL expression must evaluate to true if a device is suitable.
+	// This covers qualitative aspects of device selection.
+	//
+	// The language is as defined in
+	// https://kubernetes.io/docs/reference/using-api/cel/
+	// with several additions that are specific to device selectors.
+	//
+	// Attributes of a device are made available through a nested
```

Comment:
I think this jumps in too fast.  We can iterate on this.  It needs to be complete and digestible, without being a book.  :)  Something like:

"""
The expression's input is an object named "device", which carries the following properties:
* driver (string): the name of the driver which defines this device
* attributes (map[string]object): the device's attributes, grouped by prefix (e.g. device.attributes["example.com"] evaluates to an object with all of the attributes which were prefixed by "dra.example.com".
* capacity (map[string]object): the device's capacities, grouped by prefix.

Example: Consider a device with driver="example.com", which exposes 2 attributes named "model" and "ext.example.com/family" and which exposes 1 capacity named "modules".  This input to this expression would have the following fields:
* device.driver
* device.attributes["example.com"].model
* device.attributes["ext.example.com"].family
* device.capacity["example.com"].modules

The `device.driver` field can be used to check for a specific driver, either as a high-level precondition (i.e. you only want to consider devices from this driver) or as part of a multi-clause expression that is meant to consider devices from different drivers.

The value type of each attribute is defined by the device definition, and users who write these expressions must consult the documentation for their specific drivers.  The value type of each capacity is Quantity.

If an unknown prefix is used as a lookup in either device.attributes or device.capacity, an empty map will be returned.  Any reference to an unknown field will cause an evaluation error, which will not match the device under consideration, but may also be handled as an error.

A robust expression should check for the existence of attributes before referencing them.  

For ease of use, the cel.bind() function is enabled, and can be used to simplify expressions that access multiple attributes with the same domain.  For example:
    cel.bind(dra, device.attributes["dra.example.com"], dra.someBool && dra.anotherBool)

Reply:
Also for discussion:

"device" -> "dev"
"attributes" -> "attrs"

Reply:
> I think this jumps in too fast. We can iterate on this. It needs to be complete and digestible, without being a book. :)

There's a conflict between "usable" and "concise" here. I'm happy to take suggestions on how to strike the right balance. Ultimately, this needs to be an entire documentation page, which then begs the question of how much still should be in the API documentation.

> Also for discussion:
>
> "device" -> "dev"
> "attributes" -> "attrs"

I'm fine with the shorter names.

Reply:
Is if feasible to open a **draft** website PR with an outline of the documentation updates? If we add that PR, it could help KEP reviewers.

Reply:
Oh, and the longer names are easier for people who are slightly easier to explain for non-native speakers. Especial “dev” which abbreviates several different words used in IT.

Reply:
`device` seems short enough that we don't need to abbreviate. How familiar is `attrs` as an abbreviation for `attributes`?

`device.attrs` looks good to me.

Reply:
> Is if feasible to open a draft website PR with an outline of the documentation updates?

First I need to get this PR working and address pending review feedback.

Reply:
> I think this jumps in too fast. We can iterate on this. It needs to be complete and digestible, without being a book.

Updated. For now I am keeping "device" and "attributes". "attrs" looked odd next to "capacity".

Reply:
Continued in https://github.com/kubernetes/kubernetes/pull/125488/files#r1677650814.

Reply:
There was a good argument made for longer names. I can live with it. I guess one could `cel.bind(dev, device, ...)` or `cel.bind(attrs, device.attrs, ...)`